Xfrmi no phy dev 20250223

antonyantony opened the pull request:
do not add phy device linux xfrm interface. It is not necessary since Linux kernel 5.3.

cagney commented on programs/pluto/kernel_xfrm_interface.c:
```diff
@@ -246,7 +248,9 @@ static bool nl_newlink(const char *ipsec_if_name,
 					   "cannot find interface index for physical interface device %s", physical_if_name);
 				return false;
 			}
-			nl_addattr32(&req.n, sizeof(req.data), IFLA_XFRM_LINK, physical_if_index);
```
If `physical_if_index` isn't being used, then is the `if_nametoindex(physical_if_name)` call needed?

cagney commented on programs/pluto/ipsec_interface.c:
```diff
@@ -452,7 +453,12 @@ static struct ipsec_interface *alloc_ipsec_interface(ipsec_interface_id_t ipsec_
 			 unmap_id(ipsec_if_id));
 	passert(l < IFNAMSIZ);
-	jam_str(p->physical, sizeof(p->physical), iface->real_device_name);
+	if (kernel_xfrmi_req_phy()) {
```
Needing to specify the physical device when creating an ipsec interface seems to be a Linux only thing (going by the `ipsec-interface-00-*` tests).  Now that Linux has dropped the reqirement, why not just drop the suffix ('19 is a long time ago).

antonyantony replied:
only output change  would be  pluto log lines? "created ipsec-interface ipsec21@NONE"  "delete", while  "ip link" output would still have NONE. So I guess it is not great idea, otherwise we would have sanitize that NONE

cagney replied:
Yes, the output would change to `created ipsec-interface ipsec21` et.al.  Does `ip link` really show ipsec21@NONE!

cagney replied:
So @... can be dropped from `alloc_ipsec_interface()`  (and `ip link` just sounds broken).
Just one thing to check, if `ipsec21@eth1` is specified, does it do anything or is it ignored?

antonyantony commented:
On Sun, Feb 23, 2025 at 08:47:11AM -0800, cagney wrote:
> @cagney commented on this pull request.
> 
> 
> 
> > @@ -246,7 +248,9 @@ static bool nl_newlink(const char *ipsec_if_name,
>  					   "cannot find interface index for physical interface device %s", physical_if_name);
>  				return false;
>  			}
> -			nl_addattr32(&req.n, sizeof(req.data), IFLA_XFRM_LINK, physical_if_index);
> 
> If `physical_if_index` isn't being used, then is the `if_nametoindex(physical_if_name)` call needed?

true.This is a good catch. I moved that line only older kernel.

> -- Reply to this email directly or view it on GitHub:
> https://github.com/libreswan/libreswan/pull/2039#pullrequestreview-2635670555
> You are receiving this because you authored the thread.
> 
> Message ID: ***@***.***>

antonyantony commented:
On Sun, Feb 23, 2025 at 01:23:17PM -0800, cagney wrote:
> @cagney commented on this pull request.
> 
> 
> 
> > @@ -452,7 +453,12 @@ static struct ipsec_interface *alloc_ipsec_interface(ipsec_interface_id_t ipsec_
>  			 kernel_ops->ipsec_interface->name,
>  			 unmap_id(ipsec_if_id));
>  	passert(l < IFNAMSIZ);
> -	jam_str(p->physical, sizeof(p->physical), iface->real_device_name);
> +	if (kernel_xfrmi_req_phy()) {
> 
> Yes, the output would change to `created ipsec-interface ipsec21` et.al.  Does `ip link` really show ***@***.***!

yes, and also for vti.
what is in the console output commit is not sanitized by me.

> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/libreswan/libreswan/pull/2039#discussion_r1966891930
> You are receiving this because you authored the thread.
> 
> Message ID: ***@***.***>

cagney commented:
This:
```
		/* .interface: [id_rname@][id_vname] */
		jam_string(buf, " interface: ");
		if (c->config->ipsec_interface.enabled) {
			jam_ipsec_interface_id(buf, c->config->ipsec_interface.id);
			jam_string(buf, "@");
		}
		if (oriented(c)) {
			jam_string(buf, c->iface->real_device_name);
		}
		jam_string(buf, ";");
```
should probably be split in two with, like for vti, ipsec-interface and interface-ip and managed separated out.

cagney commented:
I guess we sort this out post 5.2

antonyantony commented:
It is very low impact as far as I can see.

cagney commented on programs/pluto/whack_connectionstatus.c:
```diff
@@ -826,7 +827,12 @@ static void show_connection_status(struct show *s, const struct connection *c)
+				jam_string(buf, none);
+			} else {
+				jam_string(buf, c->iface->real_device_name);
+			}
```
I'm fairly sure that this will show:
```
interface: NONE
```
when there's no IPsec Interface and a recent kernel.

antonyantony replied:
right. I pushed a fix.

cagney commented:
Lets push once the VMs are on f41

cagney commented:
this one has me more puzzled:
```
--- MAIN/testing/pluto/ipsec-interface-15-blocked-ikev2-xfrmi/west.console.txt
+++ OUTPUT/testing/pluto/ipsec-interface-15-blocked-ikev2-xfrmi/west.console.txt
@@ -55,7 +55,7 @@
 "westnet6-eastnet6" #4: ESP traffic information: in=0B out=0B
 west #
  ip --color=never link show ipsec1
-X: ipsec1@NONE: <NOARP,UP,LOWER_UP> mtu 1500 state UNKNOWN
+X: ipsec1@eth1: <NOARP,UP,LOWER_UP> mtu 1500 state UNKNOWN
 west #
  ip link del dev ipsec1
 west #
@@ -80,25 +80,42 @@
  ip addr add 2001:db8:0:1::251/64 dev ipsec1
 west #
  ipsec add westnet4-eastnet4
-ERROR: "westnet4-eastnet4": existing ipsec-interface ipsec1 is not valid: IFLA_XFRM_LINK attribute is zero, check 'ip -d link show dev ipsec1'
-"westnet4-eastnet4": added unoriented IKEv2 connection (neither left=192.1.2.45 nor right=192.1.2.23 match an interface)
+"westnet4-eastnet4": ipsec-interface ipsec1 linked to ? (0) and not eth1 (3)
+"westnet4-eastnet4": added IKEv2 connection
 west #
```

antonyantony commented:
I fixed it. Part of it was fixing the testcases which added "dev eth1" manually  ikev2-xfrmi-08.
I think ipsec-interface-15-blocked-ikev2-xfrmi is also fixed.
I wonder what does this comment means "# invalid device (no dev eth1)"

cagney commented:
> I wonder what does this comment means "# invalid device (no dev eth1)"

The interface created by:

```
ip link add dev ipsec1 type xfrm if_id 0x1
```

has  `dev eth1` is missing.  Pluto, seeing this, would reject the interface.  But now it should work?

antonyantony commented:
> > I wonder what does this comment means "# invalid device (no dev eth1)"
> 
> The interface created by:
> 
> ```
> ip link add dev ipsec1 type xfrm if_id 0x1
> ```
> 
> has `dev eth1` is missing. Pluto, seeing this, would reject the interface. But now it should work?

Yes it works now. I fixed tests too. Let me also check the other way around.
When creating ipsec17 externally it should allow ```dev eth1``` for completeness. I just  added ```ikev2-xfrmi-20-with-dev```

cagney commented:
Definitely merge time!

cagney commented:
Oops, looks like there's still some unintended changes:
```
testing/pluto/ikev2-xfrmi-08 failed west:iscntrl,output-different
testing/pluto/ikev2-xfrmi-10 failed west:output-different
testing/pluto/ipsec-interface-00-ipv4-ikev2-fedora-xfrm failed rise:output-different set:output-different
testing/pluto/ipsec-interface-01-ipv4-ikev2-existing-fedora failed rise:output-different set:output-different
testing/pluto/ipsec-interface-15-blocked-ikev2-xfrmi failed west:iscntrl,output-different
testing/pluto/ipsec-interface-15-order-ikev2-xfrmi failed west:output-different
testing/pluto/ipsec-interface-19-namespace-unmanaged failed west:output-different
testing/pluto/ikev2-child-ipsec-replace-ike failed west:output-different
testing/pluto/interop-ikev2-xfrmi-strongswan-01 failed east:output-different west:output-different
```

antonyantony commented:
oops. Let review the output some of it is white spaces changes! odd.
